internal/internal_task_handlers.go

```diff
@@ -484,9 +485,27 @@ func (w *workflowExecutionContextImpl) Unlock(err error) {
 			removeWorkflowContext(w.workflowInfo.WorkflowExecution.RunID)
 		} else {
 			// sticky is disabled, manually clear the workflow state.
+			if isInCache {
+				// sanity check, afaict this never happens.  but log/metric to collect evidence if it does.
+				// if it occurs it's almost certainly a logic bug (leads to double-clearing), though possibly not a damaging one.
+				// we should able to remove this after running for a while in canary / prod, our collection of users trigger quite a few edge cases.
+				w.wth.logger.DPanic("Workflow state should not have been cleared, it should be handled by cache eviction",
+					zap.String(tagWorkflowType, w.workflowInfo.WorkflowType.Name),
+					zap.String(tagWorkflowID, w.workflowInfo.WorkflowExecution.ID),
+					zap.String(tagRunID, w.workflowInfo.WorkflowExecution.RunID),
+					zap.Stack(tagPanicStack))
+				// temporary, remove after running in production for a while
+				w.wth.metricsScope.Counter("bug-clear-state")
+			}
 			w.clearState()
+			cleared = true
 		}
 	}
+	// !isInCache is the main scenario being caught here, but double-clearing is undesirable.
+	// it *might* be harmless to double-clear, but it's certainly irrational, and risks future problems.
+	if !isInCache && !cleared {
```

Comment:
Is there any reason why we are not checking a cache here directly?

Seems like it would be more straightforward to check if workflow is cached here and if so clear the state. Passing `isInCache` around add additional complexity. Or am I missing something here?

Reply:
Probably fine - I'll poke at it a bit more.  some of this is I think just leftovers from before I had a better feel for how the cache was being used / how pinning behaved.

Reply:
updated 👍 .  this is better, yeah